bug: Make internal::ConnectionImpl thread-safe.

coryan opened the pull request:
Fixes #423.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googleapis/google-cloud-cpp-spanner/406)
<!-- Reviewable:end -->

coryan commented:
For what is worth: both the tsan build and helgrind detect problems when I remove the code using the `mu_` mutex. Both pass with the current code.

devbww commented on google/cloud/spanner/internal/connection_impl.cc:
```diff
@@ -72,6 +78,14 @@ StatusOr<ConnectionImpl::SessionHolder> ConnectionImpl::GetSession() {
 }
 
+void ConnectionImpl::ReleaseSession(std::string session) {
+  if (session.empty()) {
```
Why should we make this check?

coryan commented on google/cloud/spanner/internal/connection_impl.h:
```diff
@@ -54,7 +55,10 @@ class ConnectionImpl : public Connection {
-    ~SessionHolder() { conn_->sessions_.emplace_back(std::move(session_)); }
+    ~SessionHolder() {
+      conn_->ReleaseSession(std::move(session_));
+      session_.clear();
```
Done.

devbww commented:
> What we want is to return the session exactly once. I think (but check my math) that to achieve this we need to disallow copy on SessionHolder, and make the move copy and assignment clear something to make sure the destructor only takes action once.

Do we even need move copy and assignment?  Can we just make `session_` and `conn_` const to stop anything bad from happening to a `SessionHolder`?

coryan commented on google/cloud/spanner/internal/connection_impl_test.cc:
```diff
@@ -497,6 +500,69 @@ TEST(ConnectionImplTest, RollbackSuccess) {
+  }();
+
+  auto runner = [](int thread_id, int iterations,
+                   std::shared_ptr<Connection> const& conn) {
```
Sure, but it has to be `T&` because we are calling non-const member functions.  Note that this requires a `std::ref` below because `std::async` copies every argument.

coryan commented:
> > What we want is to return the session exactly once. I think (but check my math) that to achieve this we need to disallow copy on SessionHolder, and make the move copy and assignment clear something to make sure the destructor only takes action once.
> 
> Do we even need move copy and assignment? Can we just make `session_` and `conn_` const to stop anything bad from happening to a `SessionHolder`?

I think we need at least a move constructor because we put `SessionHolder` in a `StatusOr<SessionHolder>`.

coryan commented:
@devbww PTAL.

coryan commented:
Rebased to resolve conflicts.

devbww commented:
Have you looked at the TSan failure (I haven't :-).

coryan commented:
I noticed a few minutes before you asked, so far nothing obvious, and I needed to change the TSAN build to include line numbers.

coryan commented:
Good news, it was a bug in the test. Fixed now, ran 1000 iterations of the test without error.  Rebased for good measure and will update the `.bazelrc` file in a separate PR.

devbww commented on google/cloud/spanner/internal/connection_impl_test.cc:
```diff
@@ -579,7 +579,7 @@ TEST(ConnectionImplTest, MultipleThreads) {
   std::string const session_prefix = "test-session-prefix-";
   std::string const transaction_id = "test-txn-id";
-  int session_counter = 0;
+  std::atomic<int> session_counter(0);
```
`#include <atomic>`

coryan replied:
Done.
